File: workflows/parallel-calls/code/python/parallel_calls.py
"""
Parallel Calls — Concurrent LLM calls on independent inputs, aggregated at the end.

Splits input into independent chunks, runs all LLM calls concurrently,
then combines results. Total latency ≈ max(individual call latency).

Design doc:  ../../design.md
Overview:    ../../overview.md
"""
from __future__ import annotations

from concurrent.futures import ThreadPoolExecutor, as_completed
from dataclasses import dataclass, field
from typing import Callable, Protocol


# ── LLM interface ─────────────────────────────────────────────────────────────

class LLM(Protocol):
    def generate(self, messages: list[dict]) -> str: ...


# ── Core types ────────────────────────────────────────────────────────────────

@dataclass
class BranchResult:
    index: int
    input: str
    output: str
    error: str | None = None


@dataclass
class ParallelResult:
    outputs: list[str]               # Ordered branch outputs
    aggregated: str                  # Final combined result
    errors: list[BranchResult] = field(default_factory=list)


# ── Implementation ────────────────────────────────────────────────────────────

class ParallelCalls:
    """
    Fans out to N concurrent LLM calls then aggregates the results.

    Usage:
        runner = ParallelCalls(llm, max_workers=4)
        result = runner.run(chunks, branch_prompt, aggregate_prompt)
    """

    def __init__(
        self,
        llm: LLM,
        max_workers: int = 8,
        system: str = "",
        on_error: str = "skip",       # "skip" | "raise"
    ):
        self.llm = llm
        self.max_workers = max_workers
        self.system = system
        self.on_error = on_error

    def _call_branch(self, index: int, chunk: str, prompt_template: str) -> BranchResult:
        try:
            messages: list[dict] = []
            if self.system:
                messages.append({"role": "system", "content": self.system})
            messages.append({
                "role": "user",
                "content": prompt_template.format(input=chunk),
            })
            output = self.llm.generate(messages)
            return BranchResult(index=index, input=chunk, output=output)
        except Exception as exc:
            if self.on_error == "raise":
                raise
            return BranchResult(index=index, input=chunk, output="", error=str(exc))

    def run(
        self,
        chunks: list[str],
        branch_prompt: str,           # Applied to each chunk; use {input}
        aggregate_prompt: str,        # Applied to joined outputs; use {input}
    ) -> ParallelResult:
        branch_results: list[BranchResult] = [None] * len(chunks)  # type: ignore

        with ThreadPoolExecutor(max_workers=self.max_workers) as pool:
            futures = {
                pool.submit(self._call_branch, i, chunk, branch_prompt): i
                for i, chunk in enumerate(chunks)
            }
            for future in as_completed(futures):
                result = future.result()
                branch_results[result.index] = result

        errors = [r for r in branch_results if r.error]
        outputs = [r.output for r in branch_results if not r.error]

        # Aggregate all branch outputs into one final call
        combined = "\n\n---\n\n".join(
            f"[Part {r.index + 1}]\n{r.output}"
            for r in branch_results if not r.error
        )
        agg_messages: list[dict] = []
        if self.system:
            agg_messages.append({"role": "system", "content": self.system})
        agg_messages.append({
            "role": "user",
            "content": aggregate_prompt.format(input=combined),
        })
        aggregated = self.llm.generate(agg_messages)

        return ParallelResult(outputs=outputs, aggregated=aggregated, errors=errors)


# ── Example ───────────────────────────────────────────────────────────────────

if __name__ == "__main__":
    import time

    class MockLLM:
        def generate(self, messages: list[dict]) -> str:
            time.sleep(0.05)  # Simulate network latency
            return f"[analysis of: {messages[-1]['content'][:40]}]"

    runner = ParallelCalls(llm=MockLLM(), max_workers=4)

    # Evaluate 4 independent document sections in parallel
    sections = [
        "Section 1: Market analysis shows growth in Q3...",
        "Section 2: Technical architecture uses microservices...",
        "Section 3: Financial projections indicate 20% YoY...",
        "Section 4: Risk factors include regulatory changes...",
    ]

    result = runner.run(
        chunks=sections,
        branch_prompt="Summarize this section in one sentence:\n\n{input}",
        aggregate_prompt="Combine these section summaries into an executive overview:\n\n{input}",
    )

    print(f"Branches completed: {len(result.outputs)}")
    print(f"Errors: {len(result.errors)}")
    print(f"\nAggregated:\n{result.aggregated}")

File: patterns/parallel-calls/code/typescript/vercel-ai-sdk/parallel-calls.ts
/**
 * Parallel Calls — Vercel AI SDK variant.
 *
 * Pattern: Parallel Calls (fan out N independent LLM calls on chunks of
 *   input, then aggregate the outputs in a final call). Total wall-clock
 *   approximates the slowest branch.
 * Framework: Vercel AI SDK (ai ^4.0.0) + @ai-sdk/anthropic (^1.0.0).
 * Idioms: Promise.allSettled() for the fan-out so one branch failure
 *   doesn't cancel the others; one generateText() per branch + one for
 *   the aggregate.
 * Design doc: ../../../design.md (the framework-agnostic
 *   ../../python/parallel_calls.py runs the same four-section smoke).
 *
 * Install:  pnpm add ai @ai-sdk/anthropic
 *           (or via tsx-on-demand: npx tsx parallel-calls.ts)
 * Run:      ANTHROPIC_API_KEY=... npx tsx parallel-calls.ts
 *
 * Note: ESM only.
 * tsconfig snippet: { "module": "ESNext", "moduleResolution": "Bundler", "target": "ES2022" }
 */

import { anthropic } from "@ai-sdk/anthropic";
import { generateText, type LanguageModelV1 } from "ai";

// ── Types ───────────────────────────────────────────────────────────────────

interface BranchResult {
  index: number;
  input: string;
  output: string;
  error: string | null;
}

interface ParallelResult {
  outputs: string[];
  aggregated: string;
  errors: BranchResult[];
}

// ── Driver ──────────────────────────────────────────────────────────────────

class ParallelCalls {
  constructor(
    private readonly system = "",
    private readonly model: LanguageModelV1 = anthropic("claude-haiku-4-5"),
  ) {}

  async run(
    chunks: string[],
    branchPrompt: string,
    aggregatePrompt: string,
  ): Promise<ParallelResult> {
    const settled = await Promise.allSettled(
      chunks.map((chunk, i) => this.callBranch(i, chunk, branchPrompt)),
    );

    const branchResults: BranchResult[] = settled.map((s, i) =>
      s.status === "fulfilled"
        ? s.value
        : { index: i, input: chunks[i], output: "", error: String(s.reason) },
    );

    const errors = branchResults.filter((r) => r.error !== null);
    const successes = branchResults.filter((r) => r.error === null);

    const combined = successes
      .map((r) => `[Part ${r.index + 1}]\n${r.output}`)
      .join("\n\n---\n\n");
    const aggregated = await generateText({
      model: this.model,
      system: this.system || undefined,
      prompt: aggregatePrompt.replace("{input}", combined),
    });

    return {
      outputs: successes.map((r) => r.output),
      aggregated: aggregated.text,
      errors,
    };
  }

  private async callBranch(
    index: number,
    chunk: string,
    branchPrompt: string,
  ): Promise<BranchResult> {
    const result = await generateText({
      model: this.model,
      system: this.system || undefined,
      prompt: branchPrompt.replace("{input}", chunk),
    });
    return { index, input: chunk, output: result.text, error: null };
  }
}

// ── Smoke runner ────────────────────────────────────────────────────────────

async function main(): Promise<void> {
  if (!process.env.ANTHROPIC_API_KEY) {
    console.log("Skipping smoke run — set ANTHROPIC_API_KEY to exercise the fan-out.");
    return;
  }

  const runner = new ParallelCalls();
  const sections = [
    "Section 1: Market analysis shows growth in Q3...",
    "Section 2: Technical architecture uses microservices...",
    "Section 3: Financial projections indicate 20% YoY...",
    "Section 4: Risk factors include regulatory changes...",
  ];

  const result = await runner.run(
    sections,
    "Summarize this section in one sentence:\n\n{input}",
    "Combine these section summaries into an executive overview:\n\n{input}",
  );

  console.log(`Branches completed: ${result.outputs.length}`);
  console.log(`Errors: ${result.errors.length}`);
  console.log(`\nAggregated:\n${result.aggregated}`);
}

if (import.meta.url === `file://${process.argv[1]}`) {
  main().catch((err) => {
    console.error(err);
    process.exit(1);
  });
}

export { ParallelCalls };
export type { BranchResult, ParallelResult };
